crates/sui-core/src/rest_index.rs

```diff
@@ -727,3 +621,90 @@ fn try_create_coin_index_info(object: &Object) -> Option<(CoinIndexKey, CoinInde
+
+struct RestParLiveObjectSetIndexer<'a> {
+    tables: &'a IndexStoreTables,
+    coin_index: &'a Mutex<HashMap<CoinIndexKey, CoinIndexInfo>>,
+    epoch_store: &'a AuthorityPerEpochStore,
+    package_store: &'a Arc<dyn BackingPackageStore + Send + Sync>,
+}
+
+struct RestLiveObjectIndexer<'a> {
+    tables: &'a IndexStoreTables,
+    batch: typed_store::rocks::DBBatch,
+    coin_index: &'a Mutex<HashMap<CoinIndexKey, CoinIndexInfo>>,
+    resolver: Box<dyn LayoutResolver + 'a>,
+}
+
+impl<'a> ParMakeLiveObjectIndexer for RestParLiveObjectSetIndexer<'a> {
+    type ObjectIndexer = RestLiveObjectIndexer<'a>;
+
+    fn make_live_object_indexer(&self) -> Self::ObjectIndexer {
+        RestLiveObjectIndexer {
+            tables: self.tables,
+            batch: self.tables.owner.batch(),
+            coin_index: self.coin_index,
+            resolver: self
+                .epoch_store
+                .executor()
+                .type_layout_resolver(Box::new(self.package_store)),
+        }
+    }
+}
+
+impl<'a> LiveObjectIndexer for RestLiveObjectIndexer<'a> {
+    fn index_object(&mut self, object: Object) -> Result<(), StorageError> {
+        match object.owner {
+            // Owner Index
+            Owner::AddressOwner(owner) => {
+                let owner_key = OwnerIndexKey::new(owner, object.id());
+                let owner_info = OwnerIndexInfo::new(&object);
+                self.batch
+                    .insert_batch(&self.tables.owner, [(owner_key, owner_info)])?;
+            }
+
+            // Dynamic Field Index
+            Owner::ObjectOwner(parent) => {
+                if let Some(field_info) =
+                    try_create_dynamic_field_info(&object, self.resolver.as_mut())?
+                {
+                    let field_key = DynamicFieldKey::new(parent, object.id());
+
+                    self.batch
+                        .insert_batch(&self.tables.dynamic_field, [(field_key, field_info)])?;
+                }
+            }
+
+            Owner::Shared { .. } | Owner::Immutable => {}
+        }
+
+        // Look for CoinMetadata<T> and TreasuryCap<T> objects
+        if let Some((key, value)) = try_create_coin_index_info(&object) {
+            use std::collections::hash_map::Entry;
+
+            match self.coin_index.lock().unwrap().entry(key) {
+                Entry::Occupied(o) => {
+                    let (key, v) = o.remove_entry();
+                    let value = value.merge(v);
+                    self.batch.insert_batch(&self.tables.coin, [(key, value)])?;
+                }
+                Entry::Vacant(v) => {
+                    v.insert(value);
+                }
+            }
+        }
+
+        // If the batch size grows to greater that 256MB then write out to the DB so that the
+        // data we need to hold in memory doesn't grown unbounded.
+        if self.batch.size_in_bytes() >= 1 << 28 {
+            std::mem::replace(&mut self.batch, self.tables.owner.batch()).write()?;
```

Comment:
rest indexes are actually already resistant to crashes since there is a column family used verifying initialization status that is cleared when the process starts initializing and set once its completed.